libs/wire-api/src/Wire/API/User.hs

```diff
@@ -845,22 +850,33 @@ modelUserUpdate = Doc.defineModel "UserUpdate" $ do
     Doc.description "Accent colour ID"
     Doc.optional
 
-instance ToJSON UserUpdate where
-  toJSON u =
-    A.object $
-      "name" A..= uupName u
-        # "picture" A..= uupPict u
-        # "assets" A..= uupAssets u
-        # "accent_id" A..= uupAccentId u
-        # []
-
-instance FromJSON UserUpdate where
-  parseJSON = A.withObject "UserUpdate" $ \o ->
-    UserUpdate
-      <$> o A..:? "name"
-      <*> o A..:? "picture"
-      <*> o A..:? "assets"
-      <*> o A..:? "accent_id"
+instance ToSchema UserUpdate where
+  schema =
+    object "UserUpdate" $
+      UserUpdate
+        <$> uupName .= maybe_ (optField "name" schema)
+        <*> uupPict .= maybe_ (optField "picture" schema)
+        <*> uupAssets .= maybe_ (optField "assets" (array schema))
+        <*> uupAccentId .= maybe_ (optField "accent_id" schema)
+
+data UpdateProfileError
+  = DisplayNameManagedByScim
+  | ProfileNotFound
+
+type PutSelfResponses =
+  '[ UserNotFound,
+     NameManagedByScim,
+     RespondEmpty 200 "User updated"
+   ]
+
+instance AsUnion PutSelfResponses (Maybe UpdateProfileError) where
+  toUnion (Just ProfileNotFound) = Z (I mkErrorDescription)
+  toUnion (Just DisplayNameManagedByScim) = S (Z (I mkErrorDescription))
+  toUnion Nothing = S (S (Z (I ())))
+  fromUnion (Z (I _)) = Just ProfileNotFound
+  fromUnion (S (Z (I _))) = Just DisplayNameManagedByScim
+  fromUnion (S (S (Z (I _)))) = Nothing
+  fromUnion (S (S (S x))) = case x of
```

Comment:
This should be probably easy to automate using `GenericAsUnion`.

You could also consider creating a type like `UpdateProfileResult` with 3 constructors instead, and then you could directly use `deriving via` to get the `AsUnion` instance.

Reply:
Thanks for the tip. I created `maybeToUnion` and `maybeFromUnion` (this one uses eitherFromUnion) to define these instances. Let me know if they look ok.